### PR TITLE
[RFC]Added a new function to source a string.

### LINKED ISSUE #8722: :source buffer contents without saving file
Right now, you can only `:source` files. I would be totally great if we could `:source` strings, in particular for usage in the remote API. A current workaround in the functionaltests is to write a temporary file and source that, but that comes with a set of drawbacks of course. To achieve that goal, one might look at `f_source` in `eval.c` to see if it can't be properly refactored that a possible `f_source_str` function is basically just another callee.

I want this, but I don't want to do it myself, so I'm putting up this issue ;) 

@justinmk Is this easy enough for a "good first issue", or something like that? I'd appreciated it if you could put some roughly appropriate tags on this :)

### PR DESCRIPTION
**Pull request by Siddhant085**
Fixes #8722.

Added a new function to source strings.

### CODE REVIEW

**Comment by KillTheMule** [comment]
I'm afraid the task is quite a bit bigger than that. The goal would be that the function `do_source` could be replaced by a function that reads a file into a string, and then calls `do_source_string` on it. @bfredl mentioned a ]minimal testcase](https://github.com/neovim/neovim/issues/8722#issuecomment-406791757). We also don't want to duplicate code, so you'd have to figure out how much `do_source` and `do_source_string` can share and refactor appropriately.

As long as you're working on this, add "[WIP]" to the PR title so the label can be applied.

**Comment by bfredl** [comment]
>  The goal would be that the function do_source could be replaced by a function that reads a file into a string, and then calls do_source_string on it.

I'm not sure this is necessary, `do_cmdline` is already generic for multi-line commands, it can take a callback to produce new cmdlines. The `v_list` case for `f_execute` might be an inspiration, you want something similar to `get_list_line` except that it yields the string up to the next `\n` char in a flat buffer.

**Comment by KillTheMule** [comment]
> I'm not sure this is necessary

It was meant as a theoretical possibility, to describe what it should do. As in "If you have a viml string s, and write it into a file f, the calls `:source(f)` and `:source_string(s)` should be equivalent".

**Comment by KillTheMule** [comment]
> I fail to see how my comment wouldn't make sense under that reading

I didn't say it would not make sense. I just wanted to clarify what I thought `do_source_str` should do. Obviously I'm missing some things (I did not think or know about breakpoints indeed), thanks for pointing that out. Miscommunication aside, though, we'd need to provide some guidance to @Siddhant085 about what the function really needs to do, would you speak some "final words" so it's cleared up? Thanks!

**Comment by bfredl** [comment]
In that case, my final word is to start with roughly what I said before, i e look at the v_list case for f_execute,  and do something similar to get_list_line except that it yields the string up to the next \n char in the passed String, and after that we can consider if some `do_source` logic should be shared, like profiling and/or SID (script id) maybe.

**Comment by Siddhant085** [comment]
Ok to get the test case to work I am creating a temp file inside do_source_str and then passing it to do_source. The test case is passing. I didn't see your comments before implementing this. I will look into what @bfredl mentioned.

**Comment by Siddhant085** [comment]
Using do_cmdline to execute multiline commands. Implemented `get_str_line`as a call back function to return string upto the next `\n`. [Minimal test case](https://github.com/neovim/neovim/issues/8722#issuecomment-406791757) is passing.

**Comment by KillTheMule** [comment]
Can you write tests? If not I'll provide some guidance, just let me know :)

**Comment by Siddhant085** [comment]
I don't know what kind of corner cases it should handle. Some guidance would be helpful.

**Comment by Siddhant085** [comment]
I have added a few test cases you mentioned. I tried testing for latin characters as done [here](https://github.com/neovim/neovim/blob/master/src/nvim/testdir/test_source_utf8.vim#L21), however I am unable to print unicode characters in lua. I tested it manually and its working.

**Comment by Siddhant085** [comment]
@KillTheMule and @bfredl Is everything fine or should I do something else?

**Comment by KillTheMule** [comment]
> I tested it manually and its working.

That's no substitute. Testing isn't so much about making sure it works now, rather to ensure nobody breaks it in the future.

What's the problem with utf-8 and lua? Couldn't you just c&p some utf8-chars from, say, [here](https://github.com/neovim/neovim/blob/master/test/functional/options/fillchars_spec.lua#L48) and use them?

Other than that, the linter is failing. Run `make lint` locally and fix up the issues.

One of your tests fail on windows, see [here](https://ci.appveyor.com/project/neovim/neovim/build/22022/job/v5h624eunoufeyjn#L14061). If I were you, I'd not make that test use a temporary file, but directly assert the buffer state.

Besides that, I assume the tests are insufficient (which is of course at least partially because I don't understand this very well, and I'd rather err on the side of more tests). I'd suggest the following

- Use something that contains loops, and in particular more code after the loops
- Try defining a function, and use it in another `source`
- What happens if you use `<SID>` or `<Plug>` in such a string?
- Does recursive source work properly?
- Do autocommands trigger properly?
- How about `s:a` variables?

Other people might think of more tests to do. I guess changing that status to [RFC] instead of [WIP] will attract a wild dev who would chime in :)

File: src/nvim/ex_cmds2.c [hunk 1]
```diff
@@ -2788,6 +2788,43 @@ static FILE *fopen_noinh_readbin(char *filename)
...
+  while (!(p->buf[i] == '\n' || p->buf[i] == '\0')) {
+    i++;
+  }
+  char buf[2046];
+  char *dst;
+  dst = xstpncpy(buf, (char *)p->buf+p->pointer, i - p->pointer);
+  if ((uint32_t)(dst - buf) != i - p->pointer) {
+    smsg(_("nvim_source error parsing command %s"), p->buf);
+  }
+  buf[i-p->pointer]='\0';
```
Comment (justinmk) [inline src/nvim/ex_cmds2.c]: isn't this `buf[0] = '\0'` ?

File: src/nvim/ex_cmds2.c [hunk 2]
```diff
@@ -2788,6 +2788,43 @@ static FILE *fopen_noinh_readbin(char *filename)
   return fdopen(fd_tmp, READBIN);
 }
 
+typedef struct {
+  char_u *buf;
+  uint32_t pointer;
```
Comment (justinmk) [inline src/nvim/ex_cmds2.c]: I think we would normally call this `offset`, with type `size_t`, based on how it's used.

File: src/nvim/ex_cmds2.c [hunk 3]
```diff
@@ -2788,6 +2788,43 @@ static FILE *fopen_noinh_readbin(char *filename)
...
+} GetStrLineCookie;
+
+static char_u *get_str_line(int c, void *cookie, int ident)
+{
+  GetStrLineCookie *p = (GetStrLineCookie *)cookie;
```
Comment (justinmk) [inline src/nvim/ex_cmds2.c]: we never cast `void *` pointers (C doesn't require it, and it's a source of mistakes).

File: test/functional/api/vim_spec.lua [hunk 4]
```diff
@@ -34,6 +34,84 @@ describe('API', function()
...
+    end)
+
+	it('nvim_source autocommands work', function()
+	  nvim('source','autocmd BufCreate * :let x1 = "Hello"')
+	  local fname = helpers.tmpname()
+	  nvim('command', 'new '..fname)
+	  eq(nvim('command_output','echo x1'),"Hello")
+	end)
+
+	it('nvim_source: recurssive source works', function()
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua]: "recursive"

File: src/nvim/ex_cmds2.c [hunk 5]
```diff
@@ -2788,6 +2788,43 @@ static FILE *fopen_noinh_readbin(char *filename)
...
+
+int do_source_str(char_u *cmd)
+{
+  int retval;
+  GetStrLineCookie cookie = {
+    .buf = cmd,
+    .pointer = 0,
+  };
+  retval = do_cmdline(NULL, get_str_line, (void *)&cookie,
+                      DOCMD_VERBOSE|DOCMD_NOWAIT);
```
Comment (justinmk) [inline src/nvim/ex_cmds2.c]: why `DOCMD_VERBOSE`?

File: test/functional/api/vim_spec.lua [hunk 6]
```diff
@@ -34,6 +34,84 @@ describe('API', function()
                request, nil)
   end)
 
+  describe('nvim_source', function()
+	it('works with a one-liner', function()
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua]: please indent the it() cases in the describe() block, as in the rest of the file.

File: test/functional/api/vim_spec.lua [hunk 7]
```diff
@@ -34,6 +34,84 @@ describe('API', function()
...
+	  eq(nvim('command_output','echo x1'),"Hello")
+	end)
+
+	it('nvim_source: recurssive source works', function()
+	  local fname = helpers.tmpname()
+	  nvim('command', 'new')
+	  nvim('command','edit '..fname)
+	  nvim('command','normal ilet x1 = "a"')
+	  nvim('command','w')
+	  nvim('source','source '..fname)
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua]: This is maybe a good test to have , but the recursive case looks more like:

    nvim('source', 'call nvim_source(...)'))

**Comment by justinmk** [comment]
This is a good start!

As @KillTheMule hinted, we do need to assign `current_SID` to something. Not sure what. Maybe 999999, for now.

File: src/nvim/ex_cmds2.c (hunk 1 above)
Comment (Siddhant085) [reply src/nvim/ex_cmds2.c]: We have a loop where we increment i at lines 2803-2805

**Comment by Siddhant085** [comment]
I tried setting `current_SID` to 999999. It gives seg fault at `SCRIPT_ITEM(current_SID).sn_pr_force` as 999999 exceeds the array size.

**Comment by justinmk** [comment]
Well, the idea is that 999999 would be a special value. `SCRIPT_ITEM()` and perhaps related logic would need to special-case that value ...

File: src/nvim/api/vim.c [hunk 8]
```diff
@@ -46,6 +47,14 @@
 # include "api/vim.c.generated.h"
 #endif
 
+void nvim_source(String command, Error *err)
+  FUNC_API_SINCE(5)
+{
+    try_start();
+    do_source_str((char_u *)command.data);
+    update_screen(VALID);
```
Comment (bfredl) [inline src/nvim/api/vim.c]: No need to invoke `update_screen`, it will be done after processing events (I will remove it from `nvim_command').

**Comment by justinmk** [comment]
continued in https://github.com/neovim/neovim/pull/11159

**Pull request closed without merging**